locator/abstract_replication_strategy.cc

```diff
@@ -19,6 +19,26 @@
 
 namespace locator {
 
+static endpoint_set resolve_endpoints(const host_id_set host_ids, const token_metadata2& tm) {
+    endpoint_set result{};
+    result.reserve(host_ids.size());
+    for (const auto& host_id: host_ids) {
+        // Empty host_id is used as a marker for local address.
+        // Two reasons for this hack:
+        //   * We need local_strategy to work before the local host_id is
+        //     loaded from the system.local table.
+        //   * The code that interacts with token_metadata.topology updates it
+        //     based on a specific workflow. In this workflow, the statuses of the
+        //     nodes go through various stages like bootstrapping, replacing, and normal.
+        //     Because of this, we can't simply insert the local host ID;
```

Comment:
I don't really understand this argument. What prevents us from inserting local host ID on every stage?

Reply:
IIUC @dawmd has this problem when trying to migrate `hints_manager` to be host-ID based: some code that runs concurrently with those early stages, needs to translate the local node's IP/ID back and forth. (It's on the storage_proxy/hints_manager boundary IIRC)

And the first couple topology commands sent when a node joins the cluster, do not insert that node's IP/ID entry into `locator::topology`. So the translations are failing.

This is probably because at some point someone decided that `locator::topology` will only contain entries for NORMAL nodes (?) but who says we cannot change this decision when we need to?

Or at least we can make `locator::topology` be different across nodes in a deterministic way: each node will insert extra entry to `locator::topology` corresponding to itself, even if it's just bootstrapping; even if the global topology state does not insert it yet. I see no problem with such approach.

Otherwise all places which need to do such translation would have to do a similar hack like what you did here.